File: funtime/lib/lights-draw.js
/* ============================================================
   lights-draw.js — everything you can see in Lights Out

   Black and white: a light that is ON is a black square with a white ring,
   and a light that is OFF is white with a thin grey border.
   ============================================================ */

const CELL_SIZE = 64;
const CELL_GAP = 6;
const BOARD_MARGIN = 10;

const COLOR_INK = '#111111';
const COLOR_PAPER = '#ffffff';
const COLOR_FAINT = '#cccccc';

/** boardPixelSize — how big the square canvas must be. */
function boardPixelSize() {
    return BOARD_MARGIN * 2 + GRID_SIZE * CELL_SIZE + (GRID_SIZE - 1) * CELL_GAP;
}

/** cellLeft / cellTop — where one square sits on the canvas. */
function cellLeft(column) {
    return BOARD_MARGIN + column * (CELL_SIZE + CELL_GAP);
}

function cellTop(row) {
    return BOARD_MARGIN + row * (CELL_SIZE + CELL_GAP);
}

/**
 * squareAtPixel — which square did the player click on?
 * INPUT: x, y — canvas pixels. OUTPUT: { column, row }, or null for a gap.
 */
function squareAtPixel(x, y) {
    const column = Math.floor((x - BOARD_MARGIN) / (CELL_SIZE + CELL_GAP));
    const row = Math.floor((y - BOARD_MARGIN) / (CELL_SIZE + CELL_GAP));
    if (!isOnBoard(column, row)) {
        return null;
    }
    if (x - cellLeft(column) > CELL_SIZE || y - cellTop(row) > CELL_SIZE) {
        return null;
    }
    return { column: column, row: row };
}

/** clearCanvas — paint the whole canvas one flat colour. */
function clearCanvas(ctx, width, height, color) {
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, width, height);
}

/**
 * drawLight — one square, lit or dark.
 * INPUT: ctx, isOn, column, row. OUTPUT: nothing.
 * ALGORITHM: a lit square is solid black with a white ring in the middle; an
 *            unlit one is white with a thin grey border.
 */
function drawLight(ctx, isOn, column, row) {
    const x = cellLeft(column);
    const y = cellTop(row);

    if (isOn) {
        ctx.fillStyle = COLOR_INK;
        ctx.fillRect(x, y, CELL_SIZE, CELL_SIZE);
        ctx.strokeStyle = COLOR_PAPER;
        ctx.lineWidth = 4;
        ctx.beginPath();
        ctx.arc(x + CELL_SIZE / 2, y + CELL_SIZE / 2, CELL_SIZE / 4, 0, Math.PI * 2);
        ctx.stroke();
    } else {
        ctx.fillStyle = COLOR_PAPER;
        ctx.fillRect(x, y, CELL_SIZE, CELL_SIZE);
        ctx.strokeStyle = COLOR_FAINT;
        ctx.lineWidth = 2;
        ctx.strokeRect(x + 1, y + 1, CELL_SIZE - 2, CELL_SIZE - 2);
    }
}

/** drawCursor — the dashed box showing where the keyboard cursor is. */
function drawCursor(ctx, cursor) {
    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 4;
    ctx.setLineDash([8, 6]);
    ctx.strokeRect(cellLeft(cursor.column) - 4, cellTop(cursor.row) - 4, CELL_SIZE + 8, CELL_SIZE + 8);
    ctx.setLineDash([]);
}

/** drawMessage — big centred words across the board. */
function drawMessage(ctx, width, height, title, subtitle) {
    ctx.fillStyle = 'rgba(255, 255, 255, 0.9)';
    ctx.fillRect(0, 0, width, height);
    ctx.fillStyle = COLOR_INK;
    ctx.textAlign = 'center';

    const scale = width / 370;
    const titleSize = Math.max(14, Math.round(32 * scale));
    const subtitleSize = Math.max(9, Math.round(15 * scale));

    ctx.font = 'bold ' + titleSize + 'px monospace';
    ctx.fillText(title, width / 2, height / 2 - 8);
    ctx.font = subtitleSize + 'px monospace';
    ctx.fillText(subtitle, width / 2, height / 2 + titleSize * 0.8);
}

/** renderGame — draw one complete frame. */
function renderGame(ctx, state) {
    const size = boardPixelSize();
    clearCanvas(ctx, size, size, COLOR_PAPER);

    for (let row = 0; row < GRID_SIZE; row++) {
        for (let column = 0; column < GRID_SIZE; column++) {
            drawLight(ctx, state.lights[lightIndex(column, row)], column, row);
        }
    }

    if (!state.isOver) {
        drawCursor(ctx, state.cursor);
    }

    if (state.isOver) {
        drawMessage(ctx, size, size, 'ALL OUT!', 'Solved in ' + state.moves + ' presses');
    } else if (state.isPaused) {
        drawMessage(ctx, size, size, 'PAUSED', 'Press P or tap Play to carry on');
    }
}

File: funtime/pylib/lights_draw.py
"""lights_draw.py - everything you can see in Lights Out, in Python.

Black and white: a light that is ON is a black square with a white ring, and a
light that is OFF is white with a thin grey border.
"""

import math

from lights_rules import GRID_SIZE, light_index, is_on_board

CELL_SIZE = 64
CELL_GAP = 6
BOARD_MARGIN = 10

COLOR_INK = "#111111"
COLOR_PAPER = "#ffffff"
COLOR_FAINT = "#cccccc"


def board_pixel_size():
    """How big the square canvas must be."""
    return BOARD_MARGIN * 2 + GRID_SIZE * CELL_SIZE + (GRID_SIZE - 1) * CELL_GAP


def cell_left(column):
    """The pixel position of a square's left edge."""
    return BOARD_MARGIN + column * (CELL_SIZE + CELL_GAP)


def cell_top(row):
    """The pixel position of a square's top edge."""
    return BOARD_MARGIN + row * (CELL_SIZE + CELL_GAP)


def square_at_pixel(x, y):
    """Which square did the player click on?

    INPUT:  x, y - canvas pixels. OUTPUT: (column, row), or None for a gap.
    """
    column = int((x - BOARD_MARGIN) // (CELL_SIZE + CELL_GAP))
    row = int((y - BOARD_MARGIN) // (CELL_SIZE + CELL_GAP))
    if not is_on_board(column, row):
        return None
    if x - cell_left(column) > CELL_SIZE or y - cell_top(row) > CELL_SIZE:
        return None
    return (column, row)


def clear_canvas(ctx, width, height, color):
    """Paint the whole canvas one flat colour."""
    ctx.fillStyle = color
    ctx.fillRect(0, 0, width, height)


def draw_light(ctx, is_on, column, row):
    """One square, lit or dark."""
    x = cell_left(column)
    y = cell_top(row)

    if is_on:
        ctx.fillStyle = COLOR_INK
        ctx.fillRect(x, y, CELL_SIZE, CELL_SIZE)
        ctx.strokeStyle = COLOR_PAPER
        ctx.lineWidth = 4
        ctx.beginPath()
        ctx.arc(x + CELL_SIZE / 2, y + CELL_SIZE / 2, CELL_SIZE / 4, 0, math.pi * 2)
        ctx.stroke()
    else:
        ctx.fillStyle = COLOR_PAPER
        ctx.fillRect(x, y, CELL_SIZE, CELL_SIZE)
        ctx.strokeStyle = COLOR_FAINT
        ctx.lineWidth = 2
        ctx.strokeRect(x + 1, y + 1, CELL_SIZE - 2, CELL_SIZE - 2)


def draw_cursor(ctx, cursor):
    """The dashed box showing where the keyboard cursor is."""
    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 4
    ctx.setLineDash([8, 6])
    ctx.strokeRect(cell_left(cursor["column"]) - 4, cell_top(cursor["row"]) - 4,
                   CELL_SIZE + 8, CELL_SIZE + 8)
    ctx.setLineDash([])


def draw_message(ctx, width, height, title, subtitle):
    """Big centred words across the board."""
    ctx.fillStyle = "rgba(255, 255, 255, 0.9)"
    ctx.fillRect(0, 0, width, height)
    ctx.fillStyle = COLOR_INK
    ctx.textAlign = "center"

    scale = width / 370
    title_size = max(14, round(32 * scale))
    subtitle_size = max(9, round(15 * scale))

    ctx.font = "bold " + str(title_size) + "px monospace"
    ctx.fillText(title, width / 2, height / 2 - 8)
    ctx.font = str(subtitle_size) + "px monospace"
    ctx.fillText(subtitle, width / 2, height / 2 + title_size * 0.8)


def render_game(ctx, state):
    """Draw one complete frame."""
    size = board_pixel_size()
    clear_canvas(ctx, size, size, COLOR_PAPER)

    for row in range(GRID_SIZE):
        for column in range(GRID_SIZE):
            draw_light(ctx, state["lights"][light_index(column, row)], column, row)

    if not state["is_over"]:
        draw_cursor(ctx, state["cursor"])

    if state["is_over"]:
        draw_message(ctx, size, size, "ALL OUT!", "Solved in %d presses" % state["moves"])
    elif state["is_paused"]:
        draw_message(ctx, size, size, "PAUSED", "Press P or tap Play to carry on")
